### PR TITLE
General cleanup and update of RFC0013.

### PR DESCRIPTION
**Pull request by Fraser999**
This change is [Reviewable](https://reviewable.io/reviews/maidsafe/rfcs/101)

### CODE REVIEW

File: proposed/0013-immutable-data-naming/0013-immutable-data-naming.md [hunk 1]
```diff
@@ -13,37 +13,36 @@ This RFC outlines the system components and design for how the three immutable d
...
+The SAFE Network keeps multiple copies of a single ImmutableData chunk, not only for availability but also for security reasons.  To that end, three different types of ImmutableData have been defined: Normal, Backup and Sacrificial.  These differ only in how their name is calculated; their contents are identical.
 
-Making three different types copy bearing different but determinic name will increase the difficultiy of any possible attack, as it will be three different groups, instead of just one group, need to be tackled at the same time.
+Having three different types bearing different but deterministic names will increase the difficulty of any attack, as it will be three different groups of Vault, which need to be tackled at the same time.
 
 ## Supported Use-Cases
 
-For client perspective, the use-case keeps the same : storing or fetching immutable data.  Though client may need to be aware of the existing of different types and calculations of name.
```
Comment (dirvine) [inline proposed/0013-immutable-data-naming/0013-immutable-data-naming.md]: The name calculation is unknown to clients AFAIK it should be `data.name()` for each.

File: proposed/0013-immutable-data-naming/0013-immutable-data-naming.md [hunk 2]
```diff
@@ -13,37 +13,36 @@ This RFC outlines the system components and design for how the three immutable d
...
+The SAFE Network keeps multiple copies of a single ImmutableData chunk, not only for availability but also for security reasons.  To that end, three different types of ImmutableData have been defined: Normal, Backup and Sacrificial.  These differ only in how their name is calculated; their contents are identical.
 
-Making three different types copy bearing different but determinic name will increase the difficultiy of any possible attack, as it will be three different groups, instead of just one group, need to be tackled at the same time.
+Having three different types bearing different but deterministic names will increase the difficulty of any attack, as it will be three different groups of Vault, which need to be tackled at the same time.
 
 ## Supported Use-Cases
 
-For client perspective, the use-case keeps the same : storing or fetching immutable data.  Though client may need to be aware of the existing of different types and calculations of name. 
+From the client perspective, the use-case stays the same: storing or fetching immutable data.  However, the client may need to be aware of the existence of different types and calculation of their names.
```
Comment (maqi) [inline proposed/0013-immutable-data-naming/0013-immutable-data-naming.md]: The landing (portal) IDM can carry out a mapping (i.e. put the content of received backup/sacrificial copy into a normal), so this can hide the difference of naming calculation totally from client.

File: proposed/0013-immutable-data-naming/0013-immutable-data-naming.md [hunk 3]
```diff
@@ -13,37 +13,36 @@ This RFC outlines the system components and design for how the three immutable d
...
 
-The different immutable data type copy bearing different name, which will be handled by different DataManager group.
+Replication of the different immutable data types bearing different names, which will be handled by different DataManager groups.
 
 # Detailed design
 
 ## Overview
 
-The name of an immutalbe data copy will be based on it's type and in different hash order :
+The name of a ImmutableData chunk (`im`) will be based on its type:
```
Comment (dirvine) [inline proposed/0013-immutable-data-naming/0013-immutable-data-naming.md]: Derived from it's type. Not sure we need to mention this, or perhaps mention each type has a name() method.

File: proposed/0013-immutable-data-naming/0013-immutable-data-naming.md (hunk 1 above)
Comment (Fraser999) [inline proposed/0013-immutable-data-naming/0013-immutable-data-naming.md]: OK - I'll remove the phrase "and calculation of their names".

File: proposed/0013-immutable-data-naming/0013-immutable-data-naming.md (hunk 2 above)
Comment (Fraser999) [inline proposed/0013-immutable-data-naming/0013-immutable-data-naming.md]: Yeah - I'm going to remove the phrase "and calculation of their names".  I'll leave the sentence "However, the client may need to be aware of the existence of different types." until we agree the implementation details if that's OK?

File: proposed/0013-immutable-data-naming/0013-immutable-data-naming.md (hunk 3 above)
Comment (Fraser999) [inline proposed/0013-immutable-data-naming/0013-immutable-data-naming.md]: Do you mean to just remove this one sentence?

Comment (dirvine) [inline proposed/0013-immutable-data-naming/0013-immutable-data-naming.md]: Yes probably simpler then.

File: proposed/0013-immutable-data-naming/0013-immutable-data-naming.md [hunk 4]
```diff
@@ -13,37 +13,28 @@ This RFC outlines the system components and design for how the three immutable d
...
 
 ## Supported Use-Cases
 
-For client perspective, the use-case keeps the same : storing or fetching immutable data.  Though client may need to be aware of the existing of different types and calculations of name. 
+From the client perspective, the use-case stays the same: storing or fetching immutable data.  However, the client may need to be aware of the existence of different types.
 
 ## Expected Outcome
 
-The different immutable data type copy bearing different name, which will be handled by different DataManager group.
+Replication of the different immutable data types bearing different names, which will be handled by different DataManager groups.
```
Comment (dirvine) [inline proposed/0013-immutable-data-naming/0013-immutable-data-naming.md]: Is it worth re-iterating here the Data will go the the correct `ǸaeManager` for the data. So same as all other data. The Vault will then manage each type as per normal. It requires to do no calculation on whether it should have it etc. 

The difference will be in handling it from the main copy DM who may request backup or sacrificial copies. Then again Backup may require main or sacrificial copies.  etc. 

Sacrificial will not try and get copies unless there is space in the group and it has lost both copies.

File: proposed/0013-immutable-data-naming/0013-immutable-data-naming.md [hunk 5]
```diff
@@ -13,37 +13,28 @@ This RFC outlines the system components and design for how the three immutable d
...
-Normal : normal_name = hash(im.content), handled by DM(normal_name), 4 copies on the pmid_nodes picked up by that group (one copy on each node)
-
-Backup : backup_name = hash(normal_name), handled by DM(backup_name), 4 copies on the pmid_nodes picked up by that group (one copy on each node)
-
-Sacrificial : sacrificial_name = hash(backup_name), handled by DM(sacrificial_name), 4 copies on the pmid_nodes picked up by that group (one copy on each node)
-
-MaidManagers will charge the client issuing the put request with 4 copies of the data, as the backup and sacrificial copies are allowed to be removed from the SAFE network according to the network status.
-
-The portal DMs (DataManagers that being the closest group to narmal_name), fowards the put/get requests to the second/third order DMs (i.e. DataManagers closing to backup_name and sacrificial_name). This reduces the exposure of network data types to client side to minimum.
+The Normal DMs (DataManagers comprising closest group to the Normal name), forward client Put/Get requests to the Backup and Sacrificial DMs.  This minimises client exposure to the network data types and their management.
```
Comment (dirvine) [inline proposed/0013-immutable-data-naming/0013-immutable-data-naming.md]: Only if they cannot themselves retrieve from their own group. This is a churn related error and should be addressed there.

File: proposed/0013-immutable-data-naming/0013-immutable-data-naming.md [hunk 6]
```diff
@@ -13,75 +13,91 @@ This RFC outlines the system components and design for how the three immutable d
...
-    1. Get
+If a PN is ultimately unable to store a chunk even after removing all Sacrificial chunks, it should be marked as "bad" by the DMs and not retried as a holder for that chunk.  The Put process stops at a DM group once a copy is held on two different PNs from the close group for that chunk, or once all Vaults in the close group have been attempted and failed.
 
+### Get
 
-# Drawbacks
+DMs handling a Get request for a Normal chunk from client should send a Get request to each "good" PN concurrently.  As soon as a successful response arrives, the Client's request should be responded to.
 
-None identified, other than increased complexity of Vault, Routing and Client codebase.
+Any failure responses should cause that PN to be marked as "bad" and a replacement copy should be Put to a new PN.  If both potential PNs have failed to provide the chunk, the DMs will need to send Get requests to the DMs for one of the other types of the same chunk.  Normal will request from Backup, and if this fails from Sacrificial.  Backup will request from Normal, then Sacrificial.  Sacrificial will request from Normal, then Backup.
```
Comment (maqi) [inline proposed/0013-immutable-data-naming/0013-immutable-data-naming.md]: will this causing a loop?  normal asking for backup, backup asking for sacrificial, sacrificial asking for normal ...

File: proposed/0013-immutable-data-naming/0013-immutable-data-naming.md [hunk 7]
```diff
@@ -13,75 +13,91 @@ This RFC outlines the system components and design for how the three immutable d
...
+## Planned Work
 
-2. Fraser argued that the hash of the hash paradigm shall not be used, since this could yield the same group for the normal, backup and sacrificial copies (the probability of this increasing to 1 as the network size decreases).  He proposed to XOR the chunk ID ( hash(content) ) to yield each group. While this should be more efficient than repeated hashing, more importantly it should yield groups which don't intersect (except for tiny networks).
-Essentially the "normal" group will be XOR with 000...000 i.e. equivalent to the ID.
-The sacrificial group will be XOR with fff...fff, i.e. the furthest location in the address space. And the backup will be in the middle; i.e. XOR with 800...000.
+1. MaidManager
+    1. Block attempts to Put or Get Backup or Sacrificial chunks
 
+1. DataManager
+    1. Put Backup and Sacrificial copies when handling a client Put
```
Comment (maqi) [inline proposed/0013-immutable-data-naming/0013-immutable-data-naming.md]: 1. make accounts in IDM also record immutabledatatype

File: proposed/0013-immutable-data-naming/0013-immutable-data-naming.md [hunk 8]
```diff
@@ -13,75 +13,91 @@ This RFC outlines the system components and design for how the three immutable d
...
+    1. Handle Puts from other DMs for Backup and Sacrificial chunks
+    1. Handle failure to Get by trying to Get a different type of the same chunk
+    1. Handle successful Get response of non-Normal type where this needs converted to a Normal chunk to satisfy a client Get request
+    1. Handle failure to Put on PN by Deleting Sacrificial chunks on that Vault and retrying Put
```
Comment (maqi) [inline proposed/0013-immutable-data-naming/0013-immutable-data-naming.md]: this involves a iterate through all the account entry to figure out which sacrificial copies are stored on that PN.

File: proposed/0013-immutable-data-naming/0013-immutable-data-naming.md [hunk 9]
```diff
@@ -13,75 +13,91 @@ This RFC outlines the system components and design for how the three immutable d
...
+PNs (PmidNodes) without enough free space to handle a Put request for any type chunk from PMs should respond with failure.
 
-1. Routing
-    1. ImmutableData sanity check
+DMs handling a Put failure for Normal or Backup chunks should send Delete requests for Sacrificial chunks which the same PN holds to the PM group.  Once the responses have all arrived, the initial Put request should be retried.
 
-1. Client
-    1. Put
-    1. Get
+If a PN is ultimately unable to store a chunk even after removing all Sacrificial chunks, it should be marked as "bad" by the DMs and not retried as a holder for that chunk.  The Put process stops at a DM group once a copy is held on two different PNs from the close group for that chunk, or once all Vaults in the close group have been attempted and failed.
```
Comment (maqi) [inline proposed/0013-immutable-data-naming/0013-immutable-data-naming.md]: I think the later on chapters says PN is not going to remove Sacrificial chunks?

File: proposed/0013-immutable-data-naming/0013-immutable-data-naming.md (hunk 6 above)
Comment (Fraser999) [inline proposed/0013-immutable-data-naming/0013-immutable-data-naming.md]: Well, I did state that an assumption is that we'll never have all three groups lose the data simultaneously.  In the actual code, I think we'll need to at the very least have an error message displayed if that happens, although I'd prefer a `panic!` as always :)

**Comment by dirvine** [comment]
Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions.

---

*[proposed/0013-immutable-data-naming/0013-immutable-data-naming.md, line 55 \[r3\]](https://reviewable.io:443/reviews/maidsafe/rfcs/101#-KDXiClmd16JCaiIo6fq-r3-55:-KDYjE6dUOyRQqCyKf1Q:88108764) ([raw file](https://github.com/maidsafe/rfcs/blob/eb29cb8a2861332e029f547a8b31a33232465f1c/proposed/0013-immutable-data-naming/0013-immutable-data-naming.md#L55)):*
This is not saying the PN should remove chunks itself. Maybe we can make it clear they remove chunks only when PM tells them to. 

---

*[proposed/0013-immutable-data-naming/0013-immutable-data-naming.md, line 61 \[r3\]](https://reviewable.io:443/reviews/maidsafe/rfcs/101#-KDXiClmd16JCaiIo6fq-r3-61:-KDYjPSALP6bVwYOqRqy:557920827) ([raw file](https://github.com/maidsafe/rfcs/blob/eb29cb8a2861332e029f547a8b31a33232465f1c/proposed/0013-immutable-data-naming/0013-immutable-data-naming.md#L61)):*
Only the main DM will respond to clients.  In this case a group has lost it's chunks and requires to get another copy. This is an effect of churn

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/maidsafe/rfcs/101)*

**Comment by dirvine** [comment]
Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions.

---

*[proposed/0013-immutable-data-naming/0013-immutable-data-naming.md, line 77 \[r3\]](https://reviewable.io:443/reviews/maidsafe/rfcs/101#-KDXiClmd16JCaiIo6fq-r3-77:-KDYjpXecnQq78shtruk:-1075530446) ([raw file](https://github.com/maidsafe/rfcs/blob/eb29cb8a2861332e029f547a8b31a33232465f1c/proposed/0013-immutable-data-naming/0013-immutable-data-naming.md#L77)):*
Yes each type should be explicit in code (concrete types)

---

*[proposed/0013-immutable-data-naming/0013-immutable-data-naming.md, line 81 \[r3\]](https://reviewable.io:443/reviews/maidsafe/rfcs/101#-KDXiClmd16JCaiIo6fq-r3-81:-KDYjvm0Jau4RcqUTpC1:1264138907) ([raw file](https://github.com/maidsafe/rfcs/blob/eb29cb8a2861332e029f547a8b31a33232465f1c/proposed/0013-immutable-data-naming/0013-immutable-data-naming.md#L81)):*
Yes it will. 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/maidsafe/rfcs/101)*

File: proposed/0013-immutable-data-naming/0013-immutable-data-naming.md (hunk 7 above)
Comment (Fraser999) [inline proposed/0013-immutable-data-naming/0013-immutable-data-naming.md]: I'll add that.

File: proposed/0013-immutable-data-naming/0013-immutable-data-naming.md (hunk 8 above)
Comment (Fraser999) [inline proposed/0013-immutable-data-naming/0013-immutable-data-naming.md]: Yes

**Comment by maqi** [comment]
*[proposed/0013-immutable-data-naming/0013-immutable-data-naming.md, line 22 \[r1\]](https://reviewable.io:443/reviews/maidsafe/rfcs/101#-KDXiClmd16JCaiIo6fq-r1-22:-KDYjprbijul7QxFe6Gi:-2076420285) ([raw file](https://github.com/maidsafe/rfcs/blob/c32c93e5b9f13f6fb2bd1f201f76e7dda452d776/proposed/0013-immutable-data-naming/0013-immutable-data-naming.md#L22)):*
well, i just don't want this wording causing code complexity. if you are OK with it, I am fine then. :)

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/maidsafe/rfcs/101)*

File: proposed/0013-immutable-data-naming/0013-immutable-data-naming.md (hunk 9 above)
Comment (Fraser999) [inline proposed/0013-immutable-data-naming/0013-immutable-data-naming.md]: I'll rephrase this to make it clear it's been told to delete the chunks by the DM.

**Comment by maqi** [comment]
*[proposed/0013-immutable-data-naming/0013-immutable-data-naming.md, line 55 \[r3\]](https://reviewable.io:443/reviews/maidsafe/rfcs/101#-KDXiClmd16JCaiIo6fq-r3-55:-KDYkM_qp7fl-CjWGvZP:979901340) ([raw file](https://github.com/maidsafe/rfcs/blob/eb29cb8a2861332e029f547a8b31a33232465f1c/proposed/0013-immutable-data-naming/0013-immutable-data-naming.md#L55)):*
so far, the node status is being recorded in each account entry (per chunk).
The new wording here mean IDM need to have a node-level record of node status (Bad/Good), also does it needs to be refreshed out/in during churn?
probably means an additional bullet point in the Planned work - DataManager section

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/maidsafe/rfcs/101)*

**Comment by maqi** [comment]
*[proposed/0013-immutable-data-naming/0013-immutable-data-naming.md, line 61 \[r3\]](https://reviewable.io:443/reviews/maidsafe/rfcs/101#-KDXiClmd16JCaiIo6fq-r3-61:-KDYlJ_BqnoxdhGzEVMw:-196378759) ([raw file](https://github.com/maidsafe/rfcs/blob/eb29cb8a2861332e029f547a8b31a33232465f1c/proposed/0013-immutable-data-naming/0013-immutable-data-naming.md#L61)):*
well, i just don't want this wording causing code complexity, i.e. IDM persona needs to handle the get request/response among them for different type and preventing a loop.
if you guys are OK with it, I am fine then. :)

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/maidsafe/rfcs/101)*

**Comment by Fraser999** [comment]
Review status: 0 of 1 files reviewed at latest revision, 5 unresolved discussions.

---

*[proposed/0013-immutable-data-naming/0013-immutable-data-naming.md, line 55 \[r3\]](https://reviewable.io:443/reviews/maidsafe/rfcs/101#-KDXiClmd16JCaiIo6fq-r3-55:-KDYnKhQ3n9GREkOpuR6:1428938292) ([raw file](https://github.com/maidsafe/rfcs/blob/eb29cb8a2861332e029f547a8b31a33232465f1c/proposed/0013-immutable-data-naming/0013-immutable-data-naming.md#L55)):*
Why do we need node-level records?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/maidsafe/rfcs/101)*

**Comment by maqi** [comment]
*[proposed/0013-immutable-data-naming/0013-immutable-data-naming.md, line 55 \[r3\]](https://reviewable.io:443/reviews/maidsafe/rfcs/101#-KDXiClmd16JCaiIo6fq-r3-55:-KDYounIcnNIm93eir1z:390332544) ([raw file](https://github.com/maidsafe/rfcs/blob/eb29cb8a2861332e029f547a8b31a33232465f1c/proposed/0013-immutable-data-naming/0013-immutable-data-naming.md#L55)):*
sry, wrong reply :(

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/maidsafe/rfcs/101)*

**Comment by dirvine** [comment]
Reviewed 1 of 1 files at r3.
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/maidsafe/rfcs/101)*
